File: src/main/java/clientmodel/Result.java
package clientmodel;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import servermodel.ResultState;
import org.jetbrains.annotations.NotNull;

import javax.validation.constraints.Max;
import javax.validation.constraints.Min;

@JsonAutoDetect
public class Result {
    @JsonProperty("result")
    @NotNull ResultState resultState;

    @JsonProperty("player_points")
    @Max(30)
    @Min(3)
    int playerPoints;

    @Max(30)
    @Min(3)
    @JsonProperty("opponent_points")
    int opponentPoints;

    @JsonCreator
    public Result(@NotNull @JsonProperty("result") ResultState resultState, @JsonProperty("player_points") int playerPoints, @JsonProperty("opponent_points") int opponentPoints) {
        this.resultState = resultState;
        this.playerPoints = playerPoints;
        this.opponentPoints = opponentPoints;
    }

    @Override
    public String toString() {
        return String.format("Result: %s! Your points: %d; Opponent points: %d", resultState.toString(), playerPoints, opponentPoints);
    }
}

File: src/main/kotlin/servermodel/Result.kt
package servermodel

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import javax.validation.constraints.Max
import javax.validation.constraints.Min

@JsonAutoDetect
sealed class Result @JsonCreator constructor(
    @field:JsonProperty("player_points") @param:JsonProperty("player_points") @get:JsonProperty("player_points")
    @Min(0) @Max(31) val playerPoints: Int,
    @field:JsonProperty("opponent_points") @param:JsonProperty("opponent_points") @get:JsonProperty("opponent_points")
    @Min(0) @Max(31) val opponentPoints: Int
) {
    abstract val result: ResultState
}

@JsonAutoDetect
class Winner @JsonCreator constructor(playerPoints: Int, opponentPoints: Int) : Result(playerPoints, opponentPoints) {
    @JsonProperty("result")
    override val result = ResultState.WIN
}

@JsonAutoDetect
class Loser @JsonCreator constructor(playerPoints: Int, opponentPoints: Int) : Result(playerPoints, opponentPoints) {
    @JsonProperty("result")
    override val result = ResultState.LOSE
}

@JsonAutoDetect
class Draw @JsonCreator constructor(playerPoints: Int, opponentPoints: Int) : Result(playerPoints, opponentPoints) {
    @JsonProperty("result")
    override val result = ResultState.DRAW
}